Lecture4/Tasks1-3/Task1_3L4.java

```diff
@@ -0,0 +1,318 @@
+import predefinedExceptions.RegexException;
+import predefinedExceptions.SelectedVoucherTypeException;
+import predefinedExceptions.VoucherNotFoundException;
+import predefinedExceptions.YesNoAnswerException;
+
+public class Task1_3L4 {
+
+    private static final String REGEX_YES_NO = "[ynYN]";
+    private static final String MESSAGE_TRY_AGAIN = "Please check the entered value. ";
+    private static final String MESSAGE_SORRY = "\nSorry, we don't have such vouchers :(";
+    private static final String MESSAGE_OPS = "Ops... Something went wrong.";
+
+    private static List<Voucher> listOfVouchers = new ArrayList<>();
+    private static List<Voucher> listOfSelection = new ArrayList<>();
+
+    public static void main(String[] args) {
+        try {
+            readVouchersFromFile();
+            provideFirstOffer(getUserSelection());
+            if (!isAdvancedSearchNeeded()) {
+                selectVoucher();
+            } else {
+                searchMinimumDaysCount();
+                searchWithFood();
+                searchWithTransfer();
+                provideFinalOffer();
+            }
+        } catch (OutOfMemoryError | NullPointerException mainFlowEx) {
+            System.out.println(MESSAGE_OPS);
+            System.exit(0);
+        }
+    }
+
+    private static void readVouchersFromFile() {
+        int notDownloadedCount = 0;
+        try {
+            BufferedReader readVouchersFile = new BufferedReader(new FileReader("initVouchers.txt"));
+            String sCurrentLine;
+            while ((sCurrentLine = readVouchersFile.readLine()) != null) {
+                String[] voucherDetails = sCurrentLine.split(";");
+                try {
+                    switch (Integer.parseInt(voucherDetails[0])) {
+                        case 1:
+                            Voucher restVoucher = new RestVoucher(voucherDetails);
+                            listOfVouchers.add(restVoucher);
+                            break;
+                        case 2:
+                            Voucher cureVoucher = new CureVoucher(voucherDetails);
+                            listOfVouchers.add(cureVoucher);
+                            break;
+                        case 3:
+                            Voucher shoppingVoucher = new ShoppingVoucher(voucherDetails);
+                            listOfVouchers.add(shoppingVoucher);
+                            break;
+                        default:
+                            notDownloadedCount++;
+                    }
+                } catch (NumberFormatException
+                        | ArithmeticException
+                        | ArrayIndexOutOfBoundsException
+                        | RegexException initEx) {
+                    notDownloadedCount++;
+                }
+            }
+        } catch (IOException inputEx) {
+            inputEx.printStackTrace();
+            System.exit(0);
+        }
+        if (notDownloadedCount != 0) {
```

Comment:
82-83 lines you can put to finally block if you will face not catch exceptions

Reply:
fixed in Tasks1-4-fixed package